File: include/multimedia/libqmediaformat.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qmediaformat_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qmediaformat.html
pub const qmediaformat = struct {
    /// New constructs a new QMediaFormat object.
    ///
    ///
    pub fn New() QtC.QMediaFormat {
        return qtc.QMediaFormat_new();
    }

    /// New2 constructs a new QMediaFormat object.
    ///
    /// ``` other: QtC.QMediaFormat ```
    pub fn New2(other: ?*anyopaque) QtC.QMediaFormat {
        return qtc.QMediaFormat_new2(@ptrCast(other));
    }

    /// New3 constructs a new QMediaFormat object.
    ///
    /// ``` format: qmediaformat_enums.FileFormat ```
    pub fn New3(format: i64) QtC.QMediaFormat {
        return qtc.QMediaFormat_new3(@intCast(format));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#operator=)
    ///
    /// ``` self: QtC.QMediaFormat, other: QtC.QMediaFormat ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMediaFormat_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#swap)
    ///
    /// ``` self: QtC.QMediaFormat, other: QtC.QMediaFormat ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMediaFormat_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#fileFormat)
    ///
    /// ``` self: QtC.QMediaFormat ```
    pub fn FileFormat(self: ?*anyopaque) i64 {
        return qtc.QMediaFormat_FileFormat(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#setFileFormat)
    ///
    /// ``` self: QtC.QMediaFormat, f: qmediaformat_enums.FileFormat ```
    pub fn SetFileFormat(self: ?*anyopaque, f: i64) void {
        qtc.QMediaFormat_SetFileFormat(@ptrCast(self), @intCast(f));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#setVideoCodec)
    ///
    /// ``` self: QtC.QMediaFormat, codec: qmediaformat_enums.VideoCodec ```
    pub fn SetVideoCodec(self: ?*anyopaque, codec: i64) void {
        qtc.QMediaFormat_SetVideoCodec(@ptrCast(self), @intCast(codec));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#videoCodec)
    ///
    /// ``` self: QtC.QMediaFormat ```
    pub fn VideoCodec(self: ?*anyopaque) i64 {
        return qtc.QMediaFormat_VideoCodec(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#setAudioCodec)
    ///
    /// ``` self: QtC.QMediaFormat, codec: qmediaformat_enums.AudioCodec ```
    pub fn SetAudioCodec(self: ?*anyopaque, codec: i64) void {
        qtc.QMediaFormat_SetAudioCodec(@ptrCast(self), @intCast(codec));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#audioCodec)
    ///
    /// ``` self: QtC.QMediaFormat ```
    pub fn AudioCodec(self: ?*anyopaque) i64 {
        return qtc.QMediaFormat_AudioCodec(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#isSupported)
    ///
    /// ``` self: QtC.QMediaFormat, mode: qmediaformat_enums.ConversionMode ```
    pub fn IsSupported(self: ?*anyopaque, mode: i64) bool {
        return qtc.QMediaFormat_IsSupported(@ptrCast(self), @intCast(mode));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#mimeType)
    ///
    /// ``` self: QtC.QMediaFormat ```
    pub fn MimeType(self: ?*anyopaque) QtC.QMimeType {
        return qtc.QMediaFormat_MimeType(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#supportedFileFormats)
    ///
    /// ``` self: QtC.QMediaFormat, m: qmediaformat_enums.ConversionMode, allocator: std.mem.Allocator ```
    pub fn SupportedFileFormats(self: ?*anyopaque, m: i64, allocator: std.mem.Allocator) []i64 {
        const _arr: qtc.struct_libqt_list = qtc.QMediaFormat_SupportedFileFormats(@ptrCast(self), @intCast(m));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(qmediaformat_enums.FileFormat, _arr.len) catch @panic("qmediaformat.SupportedFileFormats: Memory allocation failed");
        const _data: [*]qmediaformat_enums.FileFormat = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |_i| {
            _ret[_i] = _data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#supportedVideoCodecs)
    ///
    /// ``` self: QtC.QMediaFormat, m: qmediaformat_enums.ConversionMode, allocator: std.mem.Allocator ```
    pub fn SupportedVideoCodecs(self: ?*anyopaque, m: i64, allocator: std.mem.Allocator) []i64 {
        const _arr: qtc.struct_libqt_list = qtc.QMediaFormat_SupportedVideoCodecs(@ptrCast(self), @intCast(m));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(qmediaformat_enums.VideoCodec, _arr.len) catch @panic("qmediaformat.SupportedVideoCodecs: Memory allocation failed");
        const _data: [*]qmediaformat_enums.VideoCodec = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |_i| {
            _ret[_i] = _data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#supportedAudioCodecs)
    ///
    /// ``` self: QtC.QMediaFormat, m: qmediaformat_enums.ConversionMode, allocator: std.mem.Allocator ```
    pub fn SupportedAudioCodecs(self: ?*anyopaque, m: i64, allocator: std.mem.Allocator) []i64 {
        const _arr: qtc.struct_libqt_list = qtc.QMediaFormat_SupportedAudioCodecs(@ptrCast(self), @intCast(m));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(qmediaformat_enums.AudioCodec, _arr.len) catch @panic("qmediaformat.SupportedAudioCodecs: Memory allocation failed");
        const _data: [*]qmediaformat_enums.AudioCodec = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |_i| {
            _ret[_i] = _data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#fileFormatName)
    ///
    /// ``` fileFormat: qmediaformat_enums.FileFormat, allocator: std.mem.Allocator ```
    pub fn FileFormatName(fileFormat: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMediaFormat_FileFormatName(@intCast(fileFormat));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qmediaformat.FileFormatName: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#audioCodecName)
    ///
    /// ``` codec: qmediaformat_enums.AudioCodec, allocator: std.mem.Allocator ```
    pub fn AudioCodecName(codec: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMediaFormat_AudioCodecName(@intCast(codec));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qmediaformat.AudioCodecName: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#videoCodecName)
    ///
    /// ``` codec: qmediaformat_enums.VideoCodec, allocator: std.mem.Allocator ```
    pub fn VideoCodecName(codec: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMediaFormat_VideoCodecName(@intCast(codec));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qmediaformat.VideoCodecName: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#fileFormatDescription)
    ///
    /// ``` fileFormat: qmediaformat_enums.FileFormat, allocator: std.mem.Allocator ```
    pub fn FileFormatDescription(fileFormat: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMediaFormat_FileFormatDescription(@intCast(fileFormat));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qmediaformat.FileFormatDescription: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#audioCodecDescription)
    ///
    /// ``` codec: qmediaformat_enums.AudioCodec, allocator: std.mem.Allocator ```
    pub fn AudioCodecDescription(codec: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMediaFormat_AudioCodecDescription(@intCast(codec));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qmediaformat.AudioCodecDescription: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#videoCodecDescription)
    ///
    /// ``` codec: qmediaformat_enums.VideoCodec, allocator: std.mem.Allocator ```
    pub fn VideoCodecDescription(codec: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMediaFormat_VideoCodecDescription(@intCast(codec));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qmediaformat.VideoCodecDescription: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#operator==)
    ///
    /// ``` self: QtC.QMediaFormat, other: QtC.QMediaFormat ```
    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QMediaFormat_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#operator!=)
    ///
    /// ``` self: QtC.QMediaFormat, other: QtC.QMediaFormat ```
    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QMediaFormat_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#resolveForEncoding)
    ///
    /// ``` self: QtC.QMediaFormat, flags: qmediaformat_enums.ResolveFlags ```
    pub fn ResolveForEncoding(self: ?*anyopaque, flags: i64) void {
        qtc.QMediaFormat_ResolveForEncoding(@ptrCast(self), @intCast(flags));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediaformat.html#dtor.QMediaFormat)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QMediaFormat ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMediaFormat_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qmediaformat.html#types
pub const enums = struct {
    pub const FileFormat = enum {
        pub const UnspecifiedFormat: i32 = -1;
        pub const WMV: i32 = 0;
        pub const AVI: i32 = 1;
        pub const Matroska: i32 = 2;
        pub const MPEG4: i32 = 3;
        pub const Ogg: i32 = 4;
        pub const QuickTime: i32 = 5;
        pub const WebM: i32 = 6;
        pub const Mpeg4Audio: i32 = 7;
        pub const AAC: i32 = 8;
        pub const WMA: i32 = 9;
        pub const MP3: i32 = 10;
        pub const FLAC: i32 = 11;
        pub const Wave: i32 = 12;
        pub const LastFileFormat: i32 = 12;
    };

    pub const AudioCodec = enum {
        pub const Unspecified: i32 = -1;
        pub const MP3: i32 = 0;
        pub const AAC: i32 = 1;
        pub const AC3: i32 = 2;
        pub const EAC3: i32 = 3;
        pub const FLAC: i32 = 4;
        pub const DolbyTrueHD: i32 = 5;
        pub const Opus: i32 = 6;
        pub const Vorbis: i32 = 7;
        pub const Wave: i32 = 8;
        pub const WMA: i32 = 9;
        pub const ALAC: i32 = 10;
        pub const LastAudioCodec: i32 = 10;
    };

    pub const VideoCodec = enum {
        pub const Unspecified: i32 = -1;
        pub const MPEG1: i32 = 0;
        pub const MPEG2: i32 = 1;
        pub const MPEG4: i32 = 2;
        pub const H264: i32 = 3;
        pub const H265: i32 = 4;
        pub const VP8: i32 = 5;
        pub const VP9: i32 = 6;
        pub const AV1: i32 = 7;
        pub const Theora: i32 = 8;
        pub const WMV: i32 = 9;
        pub const MotionJPEG: i32 = 10;
        pub const LastVideoCodec: i32 = 10;
    };

    pub const ConversionMode = enum {
        pub const Encode: i32 = 0;
        pub const Decode: i32 = 1;
    };

    pub const ResolveFlags = enum {
        pub const NoFlags: i32 = 0;
        pub const RequiresVideo: i32 = 1;
    };
};

File: src/multimedia/libqmediaformat.cpp
#include <QList>
#include <QMediaFormat>
#include <QMimeType>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qmediaformat.h>
#include "libqmediaformat.h"
#include "libqmediaformat.hxx"

QMediaFormat* QMediaFormat_new() {
    return new QMediaFormat();
}

QMediaFormat* QMediaFormat_new2(const QMediaFormat* other) {
    return new QMediaFormat(*other);
}

QMediaFormat* QMediaFormat_new3(int format) {
    return new QMediaFormat(static_cast<QMediaFormat::FileFormat>(format));
}

void QMediaFormat_OperatorAssign(QMediaFormat* self, const QMediaFormat* other) {
    self->operator=(*other);
}

void QMediaFormat_Swap(QMediaFormat* self, QMediaFormat* other) {
    self->swap(*other);
}

int QMediaFormat_FileFormat(const QMediaFormat* self) {
    return static_cast<int>(self->fileFormat());
}

void QMediaFormat_SetFileFormat(QMediaFormat* self, int f) {
    self->setFileFormat(static_cast<QMediaFormat::FileFormat>(f));
}

void QMediaFormat_SetVideoCodec(QMediaFormat* self, int codec) {
    self->setVideoCodec(static_cast<QMediaFormat::VideoCodec>(codec));
}

int QMediaFormat_VideoCodec(const QMediaFormat* self) {
    return static_cast<int>(self->videoCodec());
}

void QMediaFormat_SetAudioCodec(QMediaFormat* self, int codec) {
    self->setAudioCodec(static_cast<QMediaFormat::AudioCodec>(codec));
}

int QMediaFormat_AudioCodec(const QMediaFormat* self) {
    return static_cast<int>(self->audioCodec());
}

bool QMediaFormat_IsSupported(const QMediaFormat* self, int mode) {
    return self->isSupported(static_cast<QMediaFormat::ConversionMode>(mode));
}

QMimeType* QMediaFormat_MimeType(const QMediaFormat* self) {
    return new QMimeType(self->mimeType());
}

libqt_list /* of int */ QMediaFormat_SupportedFileFormats(QMediaFormat* self, int m) {
    QList<QMediaFormat::FileFormat> _ret = self->supportedFileFormats(static_cast<QMediaFormat::ConversionMode>(m));
    // Convert QList<> from C++ memory to manually-managed C memory
    int* _arr = static_cast<int*>(malloc(sizeof(int) * _ret.length()));
    for (size_t i = 0; i < _ret.length(); ++i) {
        _arr[i] = static_cast<int>(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.length();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of int */ QMediaFormat_SupportedVideoCodecs(QMediaFormat* self, int m) {
    QList<QMediaFormat::VideoCodec> _ret = self->supportedVideoCodecs(static_cast<QMediaFormat::ConversionMode>(m));
    // Convert QList<> from C++ memory to manually-managed C memory
    int* _arr = static_cast<int*>(malloc(sizeof(int) * _ret.length()));
    for (size_t i = 0; i < _ret.length(); ++i) {
        _arr[i] = static_cast<int>(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.length();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of int */ QMediaFormat_SupportedAudioCodecs(QMediaFormat* self, int m) {
    QList<QMediaFormat::AudioCodec> _ret = self->supportedAudioCodecs(static_cast<QMediaFormat::ConversionMode>(m));
    // Convert QList<> from C++ memory to manually-managed C memory
    int* _arr = static_cast<int*>(malloc(sizeof(int) * _ret.length()));
    for (size_t i = 0; i < _ret.length(); ++i) {
        _arr[i] = static_cast<int>(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.length();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string QMediaFormat_FileFormatName(int fileFormat) {
    QString _ret = QMediaFormat::fileFormatName(static_cast<QMediaFormat::FileFormat>(fileFormat));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QMediaFormat_AudioCodecName(int codec) {
    QString _ret = QMediaFormat::audioCodecName(static_cast<QMediaFormat::AudioCodec>(codec));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QMediaFormat_VideoCodecName(int codec) {
    QString _ret = QMediaFormat::videoCodecName(static_cast<QMediaFormat::VideoCodec>(codec));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QMediaFormat_FileFormatDescription(int fileFormat) {
    QString _ret = QMediaFormat::fileFormatDescription(static_cast<QMediaFormat::FileFormat>(fileFormat));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QMediaFormat_AudioCodecDescription(int codec) {
    QString _ret = QMediaFormat::audioCodecDescription(static_cast<QMediaFormat::AudioCodec>(codec));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QMediaFormat_VideoCodecDescription(int codec) {
    QString _ret = QMediaFormat::videoCodecDescription(static_cast<QMediaFormat::VideoCodec>(codec));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

bool QMediaFormat_OperatorEqual(const QMediaFormat* self, const QMediaFormat* other) {
    return (*self == *other);
}

bool QMediaFormat_OperatorNotEqual(const QMediaFormat* self, const QMediaFormat* other) {
    return (*self != *other);
}

void QMediaFormat_ResolveForEncoding(QMediaFormat* self, int flags) {
    self->resolveForEncoding(static_cast<QMediaFormat::ResolveFlags>(flags));
}

void QMediaFormat_Delete(QMediaFormat* self) {
    delete self;
}
